dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,607 @@
+	// a ResourceClaim that has not been allocated yet.
+	//
+	// Setting this field is optional. If unset, all nodes are candidates.
+	// It is ignored when the claim does not use a control plane controller
+	// for allocation.
+	//
+	// This is an alpha field and requires enabling the DRAControlPlaneController
+	// feature gate.
+	//
+	// +optional
+	SuitableNodes *v1.NodeSelector `json:"suitableNodes,omitempty" protobuf:"bytes,4,opt,name=suitableNodes"`
 
-	// One of selector and classCriteria must be populated.
+	// Config defines configuration parameters that apply to each request in a claim using this class.
+	// They are ignored while allocating the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Config []ConfigurationParameters `json:"config,omitempty"`
 
-	// Selector specifies a label selector used to identify classes which
-	// may be considered equivalent to this class. In other words, this
-	// allows the adminstrator to define groups of classes which may be
-	// referred to as a single class in the claim.
+	// Requirements must be satisfied by devices. Applies to all devices of
+	// a claim when the claim references the class and only to the devices
+	// in a request when referenced there.
 	//
 	// +optional
-	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+	// +listType=atomic
+	Requirements []Requirement `json:"requirements,omitempty" protobuf:"bytes,4,opt,name=requirements"`
+}
 
-	// Definition specifies the criteria to determine if a device is part
-	// of this class.
-	Definition *DeviceClassDefinition `json:"definition,omitempty"`
+// ConfigurationParameters must have one and only one field set.
+type ConfigurationParameters struct {
+	Vendor *VendorConfigurationParameters `json:"vendor,omitempty" protobuf:"bytes,1,opt,name=vendor"`
 }
 
-// DeviceClassDefinition specifies the subset of devices to consider part of
-// this class, along with device-specific configuration information for those
-// devices.
-type DeviceClassDefinition struct {
-	// Driver specifies the driver that should handle this class of devices.
-	// When a DeviceClaim uses this class, only devices published by the
-	// specified driver will be considered.
-	// +required
-	Driver string `json:driver,omitempty`
+// VendorConfigurationParameters contains configuration parameters for a driver.
+type VendorConfigurationParameters struct {
+	// DriverName is used to determine which kubelet plugin needs
+	// to be passed these configuration parameters.
+	//
+	// An admission webhook provided by the vendor could use this
+	// to decide whether it needs to validate them.
+	DriverName string `json:"driverName,omitempty" protobuf:"bytes,1,opt,name=driverName"`
+
+	// Parameters can contain arbitrary data. It is the
+	// responsibility of the vendor to handle validation and
+	// versioning.
+	Parameters runtime.RawExtension `json:"parameters,omitempty" protobuf:"bytes,2,opt,name=parameters"`
```

Comment:
Are the parameters expected to be self-identifying and versioned? or is it really just "do whatever you want" ?  The only example I can see embeds a configmap, which seems pointless if it is not actually saved as an API object?

Reply:
Expected, yes, required, no.

Reply:
Can we add that to the comments here

Reply:
I think it is in the KEP, but I can also add it here.

Side node: we seem to go back and forth between "describe what a field is" without explaining how it gets used and "describe the field and who does what with it".

Personally I am leaning towards the latter because otherwise the API description is often inscrutably without knowing where to find the additional explanations. But on several occasions I  was told to cut the extra explanations - perhaps I included to much or didn't word it well. :sweat_smile:

Reply:
Updated the comment and the example class.

I was tempted to point towards apimachinery schema and type registration, but I think that would have gotten too long. It belongs into the DRA example driver as a "best practices" example with actual code that driver developers can copy-and-paste.